### PR TITLE
Add RewardsDistributorAdmin, AutoRewardsDistributor and IRewardsDistributor

### PR DESCRIPTION
**Pull request by ElliotFriedman**
Todo:
- DAO script
- E2E tests

Done:
- Add emergency kill switch to RewardsDistributorAdmin to zero out supply or borrow speeds. This function is callable by governor and guardian
- Unit testing
 - AutoRewardsDistributor
 - RewardsDistributorAdmin
- Add passthrough functions in the RewardsDistributorAdmin function to enable all necessary calls to the RewardsDistributor function
- Added ACL to RewardsDistributorAdmin that does not leverage core to make devops and deployment process simpler and lighter weight
- Added pause functionality to both RewardsDistributorAdmin and AutoRewardsDistributor contract to stop setAutoRewardsDistribution in extenuating circumstances

### CODE REVIEW

File: contracts/staking/AutoRewardsDistributor.sol [hunk 1]
```diff
@@ -0,0 +1,60 @@
+pragma solidity ^0.8.0;
+
+import "./TribalChief.sol";
+import "./RewardsDistributorAdmin.sol";
+import "../refs/CoreRef.sol";
+import "../external/rari/IRewardsDistributor.sol";
+
+/// Controller Contract to set tribe per block in Rewards Distributor Admin on Rari
```
Comment (Joeysantoro) [inline contracts/staking/AutoRewardsDistributor.sol:8]: here and elsewhere use canonical natspec annotations

File: contracts/staking/AutoRewardsDistributor.sol [hunk 2]
```diff
@@ -0,0 +1,60 @@
+pragma solidity ^0.8.0;
+
+import "./TribalChief.sol";
```
Comment (Joeysantoro) [inline contracts/staking/AutoRewardsDistributor.sol:3]: Add a TribalChief interface and use that

File: contracts/external/rari/IRewardsDistributor.sol [hunk 3]
```diff
@@ -0,0 +1,56 @@
+pragma solidity ^0.8.0;
+
+interface IRewardsDistributor {
```
Comment (Joeysantoro) [inline contracts/external/rari/IRewardsDistributor.sol:3]: Call this IRewardsDistributorAdmin and move it a new top-level folder called `feirari` with the rest of the contracts in this PR

File: contracts/staking/AutoRewardsDistributor.sol [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import "./RewardsDistributorAdmin.sol";
```
Comment (Joeysantoro) [inline contracts/staking/AutoRewardsDistributor.sol:4]: Use the interface instead

File: contracts/staking/RewardsDistributorAdmin.sol [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+        // grant this role to all AutoRewardsDistributors so that they can call 
+        _setContractAdminRole(keccak256("REWARDS_DISTRIBUTOR_ADMIN_ROLE"));
+    }
+
+    /**
+      * @notice Begins transfer of admin rights. The newPendingAdmin must call `_acceptAdmin` to finalize the transfer.
+      * @dev Admin function to begin change of admin. The newPendingAdmin must call `_acceptAdmin` to finalize the transfer.
+      * @param newPendingAdmin New pending admin.
+      */
+    function _setPendingAdmin(address newPendingAdmin) external override onlyGovernorOrAdmin {
```
Comment (Joeysantoro) [inline contracts/staking/RewardsDistributorAdmin.sol:26]: This should be exclusively `onlyGovernor`, admin shouldn't be able to change this contract

File: contracts/staking/RewardsDistributorAdmin.sol [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+    /*** Comp Distribution ***/
+   /*** Comp Distribution Admin ***/
+
+    /**
+     * @notice Transfer COMP to the recipient
+     * @dev Note: If there is not enough COMP, we do not perform the transfer all.
+     * @param recipient The address of the recipient to transfer COMP to
+     * @param amount The amount of COMP to (possibly) transfer
+     */
+    function _grantComp(address recipient, uint amount) external override onlyGovernorOrAdmin {
```
Comment (Joeysantoro) [inline contracts/staking/RewardsDistributorAdmin.sol:47]: this should be exclusively `onlyGovernor` admin shouldn't be able to withdraw TRIBE

File: contracts/staking/AutoRewardsDistributor.sol [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+        // call out to TC, grab the current allocation points for the stakedTokenWrapper
+        (,,, uint120 allocPoints,) = tribalChief.poolInfo(tribalChiefRewardIndex);
+        //   grab the total allocation points
+        uint256 totalAllocPoints = tribalChief.totalAllocPoint();
+        //   grab the tribePerBlock
+        uint256 tribePerBlock = tribalChief.tribePerBlock();
+        // Calculate the amount per block that the Rari rewards distributor is receiving from the TC
+        uint256 compSpeed = (tribePerBlock * totalAllocPoints) / allocPoints;
+        // call out to the rewards distributor admin and set the comp supply/borrow speed to the actual current value
+        return isBorrowIncentivized
```
[the remark this answers is not in the document]
Comment (ElliotFriedman) [reply contracts/staking/AutoRewardsDistributor.sol:48]: You can't have unused expressions so I had to do something with it.

I'll move to if-else syntax.

File: contracts/staking/AutoRewardsDistributor.sol [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        isBorrowIncentivized = _isBorrowIncentivized;
+        rewardsDistributorAdmin = _rewardsDistributorAdmin;
+        tribalChief = _tribalChief;
+        tribalChiefRewardIndex = _tribalChiefRewardIndex;
+        cTokenAddress = _cTokenAddress;
+    }
+
+    /// @notice function to automatically set the rewards speed on the RewardsDistributor contract
+    /// through the RewardsDistributorAdmin
+    function setAutoRewardsDistribution() external whenNotPaused {
```
[the remark this answers is not in the document]
Comment (ElliotFriedman) [reply contracts/staking/AutoRewardsDistributor.sol:38]: Should the event emit the delta or just the new compspeed?

File: contracts/feirari/RewardsDistributorAdmin.sol [hunk 9]
```diff
@@ -23,15 +24,15 @@ contract RewardsDistributorAdmin is IRewardsDistributor, CoreRef {
...
+    function _setPendingAdmin(address newPendingAdmin) external override onlyGovernor {
         rewardsDistributorContract._setPendingAdmin(newPendingAdmin);
     }
 
     /**
       * @notice Accepts transfer of admin rights. msg.sender must be pendingAdmin
       * @dev Admin function for pending admin to accept role and update admin
       */
-    function _acceptAdmin() external override onlyGovernorOrAdmin {
+    function _acceptAdmin() public override onlyGovernorOrAdmin {
```
Comment (Joeysantoro) [inline contracts/feirari/RewardsDistributorAdmin.sol:35]: When I said public I meant remove the modifier

File: contracts/feirari/RewardsDistributorAdmin.sol [hunk 10]
```diff
@@ -1,20 +1,21 @@
...
     constructor(
         address coreAddress,
-        IRewardsDistributor _rewardsDistributorContract
+        IRewardsDistributorAdmin _rewardsDistributorContract
     ) CoreRef(coreAddress) {
         rewardsDistributorContract = _rewardsDistributorContract;
-        // Create new admin role for this contract
-        // grant this role to all AutoRewardsDistributors so that they can call 
+        /// @notice Create new admin role for this contract
+        /// grant this role to all AutoRewardsDistributors so that they can call
```
Comment (Joeysantoro) [inline contracts/feirari/RewardsDistributorAdmin.sol:18]: Is this comment a directive or a description? Should be a directive i.e. "governance needs to ..."

File: contracts/staking/AutoRewardsDistributor.sol (hunk 8 above)
[the remark this answers is not in the document]
Comment (Joeysantoro) [reply contracts/staking/AutoRewardsDistributor.sol]: just new is fine, usually event is old,new but the way you have it is fine

File: contracts/feirari/AutoRewardsDistributor.sol [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+    function _deriveRequiredCompSpeed() private view returns (uint256 compSpeed) {
+        /// @notice call the TribalChief and grab the current allocation points for the stakedTokenWrapper
+        (,,, uint120 allocPoints,) = tribalChief.poolInfo(tribalChiefRewardIndex);
+        /// @notice get the total allocation points in the TribalChief
+        uint256 totalAllocPoints = tribalChief.totalAllocPoint();
+        /// @notice get the amount of tribe being distributed on a per block basis in the TribalChief
+        uint256 tribePerBlock = tribalChief.tribePerBlock();
+        /// @notice calculate the amount per block that the Rari rewards distributor is receiving from the TribalChief
+        compSpeed = (tribePerBlock * allocPoints) / totalAllocPoints;
```
Comment (ditchfieldcaleb) [inline contracts/feirari/AutoRewardsDistributor.sol:53]: If for some reason totalAllocPoints is zero (unlikely, but possible), this will throw a divide by zero error.
Checking this first as a short-circuit should fix it.

File: contracts/feirari/RewardsDistributorAdmin.sol [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+
+import "../refs/CoreRef.sol";
+import "./IRewardsDistributorAdmin.sol";
+
+contract RewardsDistributorAdmin is IRewardsDistributorAdmin, CoreRef {
+    /// @notice rewards distributor contract address
+    IRewardsDistributorAdmin public rewardsDistributorContract;
+
+    /// @param coreAddress address of core contract
+    /// @param _rewardsDistributorContract address of the admin rewards distributor contract
```
Comment (ditchfieldcaleb) [inline contracts/feirari/RewardsDistributorAdmin.sol:11]: Constructor param is a contract, but you're calling it an address here.

File: test/unit/feirari/AutoRewardsDistributor.test.ts [hunk 13]
```diff
@@ -0,0 +1,158 @@
...
+
+const RewardsDistributorAdmin = artifacts.readArtifactSync('RewardsDistributorAdmin');
+const AutoRewardsDistributor = artifacts.readArtifactSync('AutoRewardsDistributor');
+
+describe('AutoRewardsDistributor', function () {
+  let userAddress: string;
+  let minterAddress: string;
+  let pcvControllerAddress: string;
+  let governorAddress: string;
+  const e18 = '000000000000000000';
```
Comment (ditchfieldcaleb) [inline test/unit/feirari/AutoRewardsDistributor.test.ts:20]: use `ethers.constants.WeiPerEther` instead

File: test/unit/feirari/AutoRewardsDistributor.test.ts [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+
+      impersonatedSigners[address] = await ethers.getSigner(address);
+    }
+  });
+
+  beforeEach(async function () {
+    ({ userAddress, pcvControllerAddress, governorAddress, minterAddress } = await getAddresses());
+
+
+    this.poolIndex = 1000;
```
Comment (ditchfieldcaleb) [inline test/unit/feirari/AutoRewardsDistributor.test.ts:55]: Avoid using `this` inside of mocha tests; see https://stackoverflow.com/questions/27358338/mocha-this-in-before-and-beforeeach-hooks

and my comment on Joey's pr as well:

https://github.com/fei-protocol/fei-protocol-core/pull/187#discussion_r719143546

File: test/unit/feirari/AutoRewardsDistributor.test.ts [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+    this.poolAllocPoints = 1000;
+    this.totalAllocPoint = 10000;
+    this.tribePerBlock = '75' + e18;
```
Comment (ditchfieldcaleb) [inline test/unit/feirari/AutoRewardsDistributor.test.ts:58]: Never use plus when you're trying to do string concatenation; javascript has some VERY strange edge cases for this.

Instead use backtick/string substitution notation:

`this.tribePerBlock = \` 75${e18} \``

**Review by ditchfieldcaleb — COMMENTED** [review]
Added some comments, but overall looks great so far!

File: test/unit/feirari/AutoRewardsDistributor.test.ts (hunk 14 above)
Comment (ElliotFriedman) [reply test/unit/feirari/AutoRewardsDistributor.test.ts:55]: This gets redefined before each test so we should be good.

File: contracts/feirari/AutoRewardsDistributor.sol [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+        bool _isBorrowIncentivized
+    ) CoreRef(coreAddress) {
+        isBorrowIncentivized = _isBorrowIncentivized;
+        cTokenAddress = _cTokenAddress;
+        tribalChiefRewardIndex = _tribalChiefRewardIndex;
+        rewardsDistributorAdmin = _rewardsDistributorAdmin;
+        tribalChief = _tribalChief;
+    }
+
+    function _deriveRequiredCompSpeed() private view returns (uint256 compSpeed) {
```
[the remark this answers is not in the document]
Comment (ElliotFriedman) [reply contracts/feirari/AutoRewardsDistributor.sol:45]: I can change to internal as this doesn't change state. We don't inherit this contract anywhere so it shouldn't matter.

[the remark this answers is not in the document]
Comment (Joeysantoro) [reply contracts/feirari/AutoRewardsDistributor.sol:45]: Agree

File: test/unit/feirari/AutoRewardsDistributor.test.ts (hunk 13 above)
Comment (Joeysantoro) [reply test/unit/feirari/AutoRewardsDistributor.test.ts:20]: Yes!

File: test/unit/feirari/AutoRewardsDistributor.test.ts (hunk 14 above)
Comment (ditchfieldcaleb) [reply test/unit/feirari/AutoRewardsDistributor.test.ts:40]: I really think we should avoid using `this`. Another reason not to use `this` is that there's no typing on the vars when you do it this way.

File: contracts/feirari/AutoRewardsDistributor.sol [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+
+import "./../staking/ITribalChief.sol";
+import "./RewardsDistributorAdmin.sol";
+import "../refs/CoreRef.sol";
+import "./IRewardsDistributorAdmin.sol";
+
+/// @notice Controller Contract to set tribe per block in Rewards Distributor Admin on Rari
+contract AutoRewardsDistributor is CoreRef {
+    /// @notice rewards distributor admin address
+    IRewardsDistributorAdmin public rewardsDistributorAdmin;
```
[the remark this answers is not in the document]
Comment (ElliotFriedman) [reply contracts/feirari/AutoRewardsDistributor.sol:11]: Fixed.

File: test/unit/feirari/AutoRewardsDistributor.test.ts [hunk 18]
```diff
@@ -82,8 +67,7 @@ describe('AutoRewardsDistributor', function () {
 
   describe('Init', function () {
```
Comment (Joeysantoro) [inline test/unit/feirari/AutoRewardsDistributor.test.ts:68]: This init tests a bunch of stuff that isn't the AutoRewardsDistributor. Init should test the state of the contract being tested only

File: test/unit/feirari/AutoRewardsDistributor.test.ts [hunk 19]
```diff
@@ -155,4 +199,45 @@ describe('AutoRewardsDistributor', function () {
       });
     });
   });
+
+  describe('getRewardSpeedDifference', function () {
```
Comment (Joeysantoro) [inline test/unit/feirari/AutoRewardsDistributor.test.ts:201]: Test the happy path too

File: contracts/mock/MockRewardsDistributor.sol [hunk 20]
```diff
@@ -17,23 +17,29 @@ contract MockRewardsDistributor is IRewardsDistributorAdmin, Ownable {
...
 
     /**
       * @notice Begins transfer of admin rights. The newPendingAdmin must call `_acceptAdmin` to finalize the transfer.
       * @dev Admin function to begin change of admin. The newPendingAdmin must call `_acceptAdmin` to finalize the transfer.
-      * @param newPendingAdmin New pending admin.
+      * @param _newPendingAdmin New pending admin.
       */
-    function _setPendingAdmin(address newPendingAdmin) external override onlyOwner {
-        emit SuccessSetAdmin(true);
+    function _setPendingAdmin(address _newPendingAdmin) external override onlyOwner {
```
Comment (Joeysantoro) [inline contracts/mock/MockRewardsDistributor.sol:30]: Setters in our system by convention emit both the old and new one, see https://github.com/fei-protocol/fei-protocol-core/blob/master/contracts/bondingcurve/BondingCurve.sol#L259-L265 match the code style going forward

Comment (Joeysantoro) [reply contracts/mock/MockRewardsDistributor.sol:30]: including capitalization, event names, etc

File: contracts/mock/MockRewardsDistributor.sol [hunk 21]
```diff
@@ -6,9 +6,9 @@ import "../feirari/IRewardsDistributorAdmin.sol";
 import "@openzeppelin/contracts/access/Ownable.sol";
 
 contract MockRewardsDistributor is IRewardsDistributorAdmin, Ownable {
-    event SuccessSetAdmin(bool set);
-    event SuccessAcceptPendingAdmin(bool set);
-    event SuccessGrantComp(bool set);
+    event SuccessSetAdmin(bool set, address pendingAdmin);
```
Comment (Joeysantoro) [inline contracts/mock/MockRewardsDistributor.sol:9]: the bool set provides no information, is it ever false?

Comment (ElliotFriedman) [reply contracts/mock/MockRewardsDistributor.sol:9]: No. I will remove it.

File: test/unit/feirari/RewardsDistributorAdmin.test.ts [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+    await this.rewardsDistributor.transferOwnership(this.rewardsDistributorAdmin.address);
+  });
+
+  describe('Init', function () {
+    it('rewardsDistributorContract', async function () {
+      expect(await this.rewardsDistributorAdmin.rewardsDistributorContract()).to.be.equal(this.rewardsDistributor.address);
+    });
+
+    it('compSupplySpeeds', async function () {
+      expect(await this.rewardsDistributorAdmin.compSupplySpeeds(ZERO_ADDRESS)).to.be.equal(toBN('0'));
```
Comment (Joeysantoro) [inline test/unit/feirari/RewardsDistributorAdmin.test.ts]: what is this testing exactly?

File: contracts/mock/MockRewardsDistributor.sol (hunk 20 above)
Comment (Joeysantoro) [reply contracts/mock/MockRewardsDistributor.sol:30]: Events should not be camelCase.

File: test/unit/feirari/AutoRewardsDistributor.test.ts (hunk 18 above)
Comment (ElliotFriedman) [reply test/unit/feirari/AutoRewardsDistributor.test.ts:80]: fixed.

File: test/unit/feirari/AutoRewardsDistributor.test.ts (hunk 19 above)
Comment (ElliotFriedman) [reply test/unit/feirari/AutoRewardsDistributor.test.ts:192]: Added

File: contracts/feirari/RewardsDistributorAdmin.sol [hunk 23]
```diff
@@ -0,0 +1,100 @@
...
+      * @notice Accepts transfer of admin rights. msg.sender must be pendingAdmin
+      * @dev Admin function for pending admin to accept role and update admin
+      */
+    function _acceptAdmin() external override {
+        rewardsDistributorContract._acceptAdmin();
+    }
+
+    /*** Comp Distribution ***/
+   /*** Comp Distribution Admin ***/
+
```
Comment (ElliotFriedman) [inline contracts/feirari/RewardsDistributorAdmin.sol:41]: Add method here that calls _setImplementation as this will hit a proxy contract.

File: contracts/staking/AutoRewardsDistributor.sol (hunk 2 above)
Comment (ElliotFriedman) [reply contracts/staking/AutoRewardsDistributor.sol]: Done

File: contracts/staking/AutoRewardsDistributor.sol (hunk 4 above)
Comment (ElliotFriedman) [reply contracts/staking/AutoRewardsDistributor.sol]: fixed

File: contracts/feirari/RewardsDistributorAdmin.sol (hunk 10 above)
Comment (ElliotFriedman) [reply contracts/feirari/RewardsDistributorAdmin.sol]: This is a directive. Fixed.

File: contracts/feirari/AutoRewardsDistributor.sol (hunk 11 above)
Comment (ElliotFriedman) [reply contracts/feirari/AutoRewardsDistributor.sol]: fixed

File: contracts/feirari/RewardsDistributorAdmin.sol (hunk 12 above)
Comment (ElliotFriedman) [reply contracts/feirari/RewardsDistributorAdmin.sol]: Fixed

File: test/unit/feirari/AutoRewardsDistributor.test.ts (hunk 13 above)
Comment (ElliotFriedman) [reply test/unit/feirari/AutoRewardsDistributor.test.ts]: done

File: test/unit/feirari/AutoRewardsDistributor.test.ts (hunk 14 above)
Comment (ElliotFriedman) [reply test/unit/feirari/AutoRewardsDistributor.test.ts]: fixed

File: test/unit/feirari/AutoRewardsDistributor.test.ts (hunk 15 above)
Comment (ElliotFriedman) [reply test/unit/feirari/AutoRewardsDistributor.test.ts]: fixed

File: contracts/mock/MockRewardsDistributor.sol (hunk 21 above)
Comment (ElliotFriedman) [reply contracts/mock/MockRewardsDistributor.sol]: done

File: test/unit/feirari/RewardsDistributorAdmin.test.ts (hunk 22 above)
Comment (ElliotFriedman) [reply test/unit/feirari/RewardsDistributorAdmin.test.ts]: testing that initial compSupplySpeeds are 0. I pass 0 address as this doesn't matter in the test because it is just hitting a mock.

File: contracts/feirari/RewardsDistributorAdmin.sol [hunk 24]
```diff
@@ -2,21 +2,33 @@ pragma solidity ^0.8.0;
 
 import "../refs/CoreRef.sol";
 import "./IRewardsDistributorAdmin.sol";
+import "@openzeppelin/contracts/access/AccessControlEnumerable.sol";
+
+contract RewardsDistributorAdmin is IRewardsDistributorAdmin, CoreRef, AccessControlEnumerable {
+
+    bytes32 public constant override AUTO_REWARDS_DISTRIBUTOR = keccak256("AUTO_REWARDS_DISTRIBUTOR");
```
Comment (Joeysantoro) [inline contracts/feirari/RewardsDistributorAdmin.sol:9]: call this AUTO_REWARDS_DISTRIBUTOR_ROLE

File: contracts/feirari/RewardsDistributorAdmin.sol [hunk 25]
```diff
@@ -2,21 +2,33 @@ pragma solidity ^0.8.0;
...
         address coreAddress,
-        IRewardsDistributorAdmin _rewardsDistributorContract
+        IRewardsDistributorAdmin _rewardsDistributorContract,
+        address[] memory _autoRewardDistributors
     ) CoreRef(coreAddress) {
         rewardsDistributorContract = _rewardsDistributorContract;
         /// @notice Governance should create new admin role for this contract
         /// and then grant this role to all AutoRewardsDistributors so that they can call this contract
-        _setContractAdminRole(keccak256("REWARDS_DISTRIBUTOR_ADMIN_ROLE"));
+        _setContractAdminRole(keccak256("TRIBAL_CHIEF_ADMIN_ROLE"));
```
Comment (Joeysantoro) [inline contracts/feirari/RewardsDistributorAdmin.sol:24]: Add a comment explaining why we reuse the TribalChief one

File: contracts/feirari/RewardsDistributorAdmin.sol [hunk 26]
```diff
@@ -2,21 +2,33 @@ pragma solidity ^0.8.0;
...
+        /// assign the admin role of AUTO_REWARDS_DISTRIBUTOR to DEFAULT_ADMIN_ROLE
```
Comment (Joeysantoro) [inline contracts/feirari/RewardsDistributorAdmin.sol:25]: comments that don't provide additional context (like this one) can be taken out

Comment (Joeysantoro) [reply contracts/feirari/RewardsDistributorAdmin.sol:25]: More useful would be explaining that there are two different access control systems here, one in Core and one locally. Explaining the roles and relationships probably in a top level block comment above the contract definition will be ideal

File: contracts/feirari/AutoRewardsDistributor.sol [hunk 27]
```diff
@@ -0,0 +1,102 @@
...
+    }
+
+    /// @notice function to automatically set the rewards speed on the RewardsDistributor contract
+    /// through the RewardsDistributorAdmin
+    function setAutoRewardsDistribution() external whenNotPaused {
+        (uint256 compSpeed, bool updateNeeded) = getRewardSpeedDifference();
+        require(updateNeeded, "AutoRewardsDistributor: update not needed");
+
+        /// @notice call out to the rewards distributor admin and set the comp supply/borrow speed to the current value
+        /// if the contract is borrow incentivized, set that speed, else, set the supply speed
```
Comment (Joeysantoro) [inline contracts/feirari/AutoRewardsDistributor.sol:86]: line 86 redundant comment

File: contracts/feirari/AutoRewardsDistributor.sol [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+        if (isBorrowIncentivized) {
+            rewardsDistributorAdmin._setCompBorrowSpeed(cTokenAddress, compSpeed);
+        } else {
+            rewardsDistributorAdmin._setCompSupplySpeed(cTokenAddress, compSpeed);
+        }
+        emit SpeedChanged(isBorrowIncentivized, compSpeed, cTokenAddress);
```
Comment (Joeysantoro) [inline contracts/feirari/AutoRewardsDistributor.sol:92]: isborrowIncentivized is immutable state, same with cTokenAddress. These can be removed from the event. Normally events only want to emit changed/mutable state

File: contracts/feirari/AutoRewardsDistributor.sol [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+    /// @notice rewards distributor admin contract
+    IRewardsDistributorAdmin public rewardsDistributorAdmin;
+    /// @notice tribal chief contract
+    ITribalChief public tribalChief;
+    /// @notice address of the CToken this contract controls rewards for
+    address public cTokenAddress;
+    /// @notice boolean which decides the action to incentivize
+    bool public isBorrowIncentivized;
+    /// @notice reward index on tribal chief to grab this staked token wrapper's index
+    uint256 public tribalChiefRewardIndex;
```
Comment (Joeysantoro) [inline contracts/feirari/AutoRewardsDistributor.sol:18]: All of these can be immutable except rewardDistributorAdmin

File: contracts/feirari/AutoRewardsDistributor.sol [hunk 30]
```diff
@@ -0,0 +1,102 @@
...
+        }
+        emit SpeedChanged(isBorrowIncentivized, compSpeed, cTokenAddress);
+    }
+
+    /// @notice API to point to a new rewards distributor admin contract
+    /// @param _rewardsDistributorAdmin the address of the new RewardsDistributorAdmin contract
+    function setRewardsDistributorAdmin(
+        IRewardsDistributorAdmin _rewardsDistributorAdmin
+    ) external onlyGovernorOrAdmin {
+        rewardsDistributorAdmin = _rewardsDistributorAdmin;
```
Comment (Joeysantoro) [inline contracts/feirari/AutoRewardsDistributor.sol:100]: This should have an event

File: contracts/feirari/RewardsDistributorAdmin.sol [hunk 31]
```diff
@@ -0,0 +1,129 @@
...
+
+contract RewardsDistributorAdmin is IRewardsDistributorAdmin, CoreRef, AccessControlEnumerable {
+
+    bytes32 public constant override AUTO_REWARDS_DISTRIBUTOR = keccak256("AUTO_REWARDS_DISTRIBUTOR");
+
+    /// @notice rewards distributor contract
+    IRewardsDistributorAdmin public rewardsDistributorContract;
+
+    /// @param coreAddress address of core contract
+    /// @param _rewardsDistributorContract admin rewards distributor contract
```
Comment (Joeysantoro) [inline contracts/feirari/RewardsDistributorAdmin.sol:15]: add _autoRewardsDistributors to natspec

File: contracts/feirari/RewardsDistributorAdmin.sol [hunk 32]
```diff
@@ -0,0 +1,129 @@
...
+      * @notice Begins transfer of admin rights. The newPendingAdmin must call `_acceptAdmin` to finalize the transfer.
+      * @dev Admin function to begin change of admin. The newPendingAdmin must call `_acceptAdmin` to finalize the transfer.
+      * @param newPendingAdmin New pending admin.
+      */
+    function _setPendingAdmin(address newPendingAdmin) external override onlyGovernor {
+        rewardsDistributorContract._setPendingAdmin(newPendingAdmin);
+    }
+
+    /**
+      * @notice Accepts transfer of admin rights. msg.sender must be pendingAdmin
```
Comment (Joeysantoro) [inline contracts/feirari/RewardsDistributorAdmin.sol:44]: this comment is slightly inaccurate. "this" must be pendingAdmin not the sender

Comment (ElliotFriedman) [reply contracts/feirari/RewardsDistributorAdmin.sol:44]: I just copy and pasted this comment from the rari codebase as it is a passthrough.

File: contracts/feirari/RewardsDistributorAdmin.sol (hunk 31 above)
Comment (ElliotFriedman) [reply contracts/feirari/RewardsDistributorAdmin.sol:19]: added

File: contracts/feirari/AutoRewardsDistributor.sol (hunk 30 above)
Comment (ElliotFriedman) [reply contracts/feirari/AutoRewardsDistributor.sol]: added

File: contracts/feirari/AutoRewardsDistributor.sol (hunk 29 above)
Comment (ElliotFriedman) [reply contracts/feirari/AutoRewardsDistributor.sol]: added

File: contracts/feirari/AutoRewardsDistributor.sol (hunk 27 above)
Comment (ElliotFriedman) [reply contracts/feirari/AutoRewardsDistributor.sol]: removed

File: contracts/feirari/AutoRewardsDistributor.sol (hunk 28 above)
Comment (ElliotFriedman) [reply contracts/feirari/AutoRewardsDistributor.sol]: done

File: contracts/feirari/RewardsDistributorAdmin.sol (hunk 26 above)
Comment (ElliotFriedman) [reply contracts/feirari/RewardsDistributorAdmin.sol]: Removed useless comment and added context comment to top of file

File: contracts/feirari/RewardsDistributorAdmin.sol (hunk 24 above)
Comment (ElliotFriedman) [reply contracts/feirari/RewardsDistributorAdmin.sol]: done

File: contracts/feirari/RewardsDistributorAdmin.sol [hunk 33]
```diff
@@ -4,15 +4,20 @@ import "../refs/CoreRef.sol";
 import "./IRewardsDistributorAdmin.sol";
 import "@openzeppelin/contracts/access/AccessControlEnumerable.sol";
 
+/// @notice this contract has its own internal RBAC. The reason for doing this
```
Comment (Joeysantoro) [inline contracts/feirari/RewardsDistributorAdmin.sol:7]: Not familiar with this acronym. ACL is the one I am more familiar with.

Comment (ElliotFriedman) [reply contracts/feirari/RewardsDistributorAdmin.sol:7]: Role Based Access Control

File: contracts/feirari/RewardsDistributorAdmin.sol [hunk 34]
```diff
@@ -0,0 +1,135 @@
...
+      * @dev Admin function for pending admin to accept role and update admin
+      */
+    function _acceptAdmin() external override {
+        rewardsDistributorContract._acceptAdmin();
+    }
+
+    /*** Comp Distribution ***/
+   /*** Comp Distribution Admin ***/
+
+    /**
```
Comment (ElliotFriedman) [inline contracts/feirari/RewardsDistributorAdmin.sol:60]: Add method that will allow guardian or governor to zero rewards for any cToken. This will be very useful in case an emergency shutdown is needed.

File: test/unit/feirari/RewardsDistributorAdmin.test.ts [hunk 35]
```diff
@@ -13,8 +13,13 @@ const {
     constants: { ZERO_ADDRESS }
 } = testHelpers;
 
+/// TODO
```
Comment (Joeysantoro) [inline test/unit/feirari/RewardsDistributorAdmin.test.ts:16]: Whats going on with these comments?

**Review by Joeysantoro — APPROVED** [review]

File: test/unit/feirari/RewardsDistributorAdmin.test.ts (hunk 35 above)
Comment (ElliotFriedman) [reply test/unit/feirari/RewardsDistributorAdmin.test.ts]: Removed. It was for functions in RDA that I needed to test.
